File: po2db.pl
#!/usr/bin/perl

BEGIN{
    use File::Basename;
    # if(-l $0){
    # 	use lib dirname($0) .'/'. dirname(readlink $0);
    # }else {
    use lib dirname($0);
    # }
}
use strict;
use warnings;
use Encode;
use PO2DB;
use DBI;
use Carp;

my $db_file="main.sqlite";

#get the db name
if (! ($ARGV[0]=~ /\.po$/)) {
    $db_file=shift;
}
# if(-e $db_file){
#     croak "\e[0;31m$db_file is existed\e[0m\n";
# }


#get the table suffix
my $table_suffix='default';
unless( -f $ARGV[0]){
    $table_suffix=(shift);
}
$table_suffix=~s/^-++//;
# if($table_suffix=~/([\/\.-])/){
#     croak "\e[0;31millegal table suffix [$1]\e[0m\n";
# }

my @po_files=sort @ARGV;
#try to reduce the path length
my ($path_f,$path_l,$path_pre)=(@po_files[0,$#po_files],'');
if (defined $path_f) {
    while ($path_f=~/($path_pre[^\/]*\/)/) {
	my $t=$1;
	last unless($path_l=~/^$path_pre/);
	$path_pre=$t;
    }
}


$table_suffix= $path_pre if $table_suffix eq 'default';
$table_suffix=~s/\/$//;
$table_suffix=~s/.*\///;
$table_suffix=~s/'//g;

my $t1="t_$table_suffix";	#translations
my $t2="h_$table_suffix";	#headerinfo
my $i1="i_$t1";
my $i2="i_$t2";

my $id=0;


my $dbh;
$dbh= DBI->connect("dbi:SQLite:dbname=$db_file","","");
my %tables=map {$$_[0],undef} @{$dbh->selectall_arrayref("SELECT name FROM sqlite_master WHERE type='table';")};
$dbh->do('BEGIN TRANSACTION');
# $dbh->do("drop table if exists $t1");
# $dbh->do("drop table if exists $t2");
if (exists $tables{$t1}) {
    (${[reverse sort grep {/${t1}/} keys %tables]}[0])=~/(\d+)$/;
    my $i;
    if (defined $1) {
	$i=$1;
	$i++;
    } else {
	$i='0';
    }
    $dbh->do("alter table '$t1' rename to '${t1}_$i'");
    $dbh->do("alter table '$t2' rename to '${t2}_$i'");
    $dbh->do("drop index if exists '$i1'");
    $dbh->do("drop index if exists '$i2'");
    $dbh->do("create index '${i1}_$i' on '${t1}_$i' (id,msgid,msgstr,msgctxt,fuzzy,flag,pof)");
    $dbh->do("create index '${i2}_$i' on '${t2}_$i' (pof,lname,lmail,tname,tmail,charset,pforms)");
    if ($i=~/^10+$/) {
	for my $j (0..$i-1) {
	    my $j1=sprintf "%0".(log($i)/log(10))."d",$j;
	    my $j2=sprintf "%0".(log($i)/log(10)+1)."d",$j;
	    $dbh->do("alter table '${t1}_$j1' rename to '${t1}_$j2'");
	    $dbh->do("alter table '${t2}_$j1' rename to '${t2}_$j2'");
	    $dbh->do("drop index if exists '${i1}_$j1'");
	    $dbh->do("drop index if exists '${i2}_$j1'");
	    $dbh->do("create index '${i1}_$j2' on '${t1}_$j2' (id,msgid,msgstr,msgctxt,fuzzy,flag,pof)");
	    $dbh->do("create index '${i2}_$j2' on '${t2}_$j2' (pof,lname,lmail,tname,tmail,charset,pforms)");
	}
    }
    #    $dbh->do('commit');$dbh->commit;exit;
}
$dbh->do("create table '$t1' (id integer,msgid text,msgstr text,msgctxt text,fuzzy bool,flag text,pof text)");
$dbh->do("create table '$t2' (pof text,lname text,lmail text,tname text,tmail text,charset text,pforms text)");


for my $fn (@po_files) {
    my $pof=$fn;
    $pof=~s/'/''/g;
    my($trans,$trans_e,$team,$team_e,$charset,$pf)=headerinfo($fn);
    $dbh->do("insert into '$t2' values('$pof','$trans','$trans_e','$team','$team_e','$charset','$pf')");
    my($msgid,$msgstr,$msgctxt,$flag,$fuzzy);
    for (trans_array($fn)) {
	($msgid,$msgstr,$msgctxt,$flag)=@$_;
	next if $msgid eq '';
	if ($flag=~/fuzzy/) {
	    $fuzzy=1;
	    $flag=join ",",grep !/fuzzy/,split ",",$flag;
	} else {
	    $fuzzy=0;
	    $flag=~s/ +//g;
	}
	++$id;
	$dbh->do("insert into '$t1' values($id,'$msgid','$msgstr','$msgctxt',$fuzzy,'$flag','$pof');");
    }
}
$dbh->do("create index '$i1' on '$t1' (id,msgid,msgstr,msgctxt,fuzzy,flag,pof)");
$dbh->do("create index '$i2' on '$t2' (pof,lname,lmail,tname,tmail,charset,pforms)");
$dbh->do('commit');
$dbh->commit;

File: po2db.lua
#!/usr/bin/env luajit

local my_assert=assert
local push = function(arr,item)
   arr[#arr+1]=item
end

local function get_file_as_t (file)
   local content = {}
   local i = 0;
   for line in io.lines(file) do
      push(content,line)
   end
   return content
end

local get_items = function (content)
   local items = {}
   local find=string.find
   local match=string.match
   local format=string.format
   local key,item=nil,{}
   for i,line in ipairs(content) do
      if line:find([[^"]]) then
	 my_assert(key,format("tailing string in wrong position,line number: %d\nline:%s",i,line))
	 if key ~= "m-other" then
	    item[key]=item[key] .. line:match([["(.*)"]])
	 end
      elseif line:find([[^msgid%s+"]]) or line:find([[^msgid%[0%]%s+]]) then
	 key="msgid"
	 my_assert(not item[key],format("duplicate %s,line number:%d",key,i))
	 item[key]=line:match([["(.*)"]])
      elseif line:find([[^msgstr%s+"]]) or line:find([[^msgstr%[0%]%s+]]) then
	 key = "msgstr"
	 my_assert(not item[key],format("duplicate %s,line number:%d",key,i))
	 item[key]=line:match([["(.*)"]])
      elseif line:find([[^msgctxt%s+]]) then
	 key = "msgctxt"
	 my_assert(not item[key],format("duplicate %s,line number:%d",key,i))
	 item[key]=line:match([["(.*)"]])
      elseif line:find([[^#,]]) then
	 key = "flag"
	 my_assert(not item[key],format("duplicate %s,line number:%d",key,i))
	 item[key]=line:match([[^#,%s*(.*)]]):gsub("%s+","")
      elseif line:find([[^%s*$]]) then
	 if item["msgid"] and item["msgstr"] then
	    push(items,item)
	    item={}
	 end
      elseif line:find([[^#~]]) then
	 item["flag"]=nil
      elseif line:find [[^#]] then
      elseif line:find [[^m]] then
	 key = "m-other"
	 item[key]=line:match([["(.*)"]])
      else
	 my_assert(nil,format("undefined format for {%s},line number:%d",line,i))
      end
   end
   if item["msgid"] and item["msgstr"] then
      push(items,item)
      item={}
   end
   return items
end


local function headerinfo (content)
   local trans,trans_e,team,team_e,charset,pf
   local match = string.match
   local find = string.find
   for _,line in ipairs(content) do
      -- "Last-Translator: Yinghua Wang <wantinghard@gmail.com>\n"
      if(find(line,[[^"Last%-Translator]])) then
      	 trans,trans_e=match(line,[[^"Last%-Translator: *([^<]+[^ <]) *<([^>]+)>]])
      	 -- "Language-Team: Chinese (simplified) <i18n-zh@googlegroups.com>\n"
      elseif(find(line,[[^"Language%-Team:]])) then
      	 team,team_e=match(line,[[^"Language%-Team: *([^<]+[^ <]) *<([^>]+)>]])
      elseif(find(line,[[^"Content%-Type]])) then
      	 charset=match(line,[[^"Content%-Type: text/plain; charset=([^ ]+) *\n]])
      	 -- "Plural-Forms: nplurals=1; plural=0;\n"
      elseif(find(line,[[^"Plural%-Forms: ]])) then
      	 pf=match(line,[[^"Plural%-Forms: *(.+[^ ]) *\n"]])
      end
   end
   return trans,trans_e,team,team_e,charset,pf
end

local function parse_opt(arg)
   require "alt_getopt"
   local long_opts = {
      verbose = "v",
      help    = "h",
      db_file_path = "d",
      table_suffix = "t",
   }

   local optarg,optind = alt_getopt.get_opts (arg,"vhd:t:", long_opts)
   
   local verbose,help,db_file_path,table_suffix=optarg["v"],optarg["h"],optarg["d"] or "main.sqlite",optarg["t"] or "default"
   local poes={}
   for i = optind,#arg do
      push(poes,arg[i])
   end
   print (verbose,help,db_file_path,table_suffix,poes)
end

local function setting_table_and_index(con,trans_table,header_table)
   local cur
   local format = string.format
   for _,table in ipairs({trans_table,header_table}) do
      table = con:escape(table)
      cur = assert(con:execute(format("select count(name) from sqlite_master where name == '%s';",table)))
      local table_exists_p = (cur:fetch())
      cur:close()
      if(table_exists_p) then
	 local cur = assert(con:execute(format("select count(name) from sqlite_master where name like '%s%%';",table)))
	 local number_of_tables = cur:fetch()
	 cur:close()
	 assert(con:execute(format("alter table '%s' rename to '%s_%d';'",table,table,number_of_tables-1)))
      end
   end
   assert(con:execute(format("create table '%s' (id integer,msgid text,msgstr text,msgctxt text,fuzzy bool,flag text,pof text);\n",con:escape(trans_table))))
   assert(con:execute(format("create table '%s' (pof text,lname text,lmail text,tname text,tmail text,charset text,pforms text);\n",con:escape(header_table))))
end

local function main ()
   local verbose,help,db_file_path,table_suffix,poes = parse_opt(arg)
   local format=string.format
   local luasql=require("luasql.sqlite3")
   local clock=os.clock
   local tm_parsing,tm_db=0,0
   local env = assert (luasql.sqlite3())
   local con = assert (env:connect("test.sqlite"))
   local t
   setting_table_and_index(con,"t_default","h_default")
   os.exit()
   local e = function (t,i)
      if t[i] then
	 return con:escape(t[i])
      else
	 return ""
      end
   end
   local res
   res = assert (con:execute[[BEGIN TRANSACTION]])
   res = assert (con:execute[[create table 'tab' (id integer,msgid text,msgstr text,msgctxt text,fuzzy bool,flag text,pof text)]])
   for _,file in ipairs(arg) do
      tm_parsing=tm_parsing-clock()
      local content=get_file_as_t(file)
      local items=get_items(content)
      tm_parsing=tm_parsing+clock()
      file=con:escape(file)
      tm_db=tm_db-clock()
      for i,item in ipairs(items) do
	 res = assert(con:execute(format("insert into 'tab' values(%d,'%s','%s','%s',1,'aou','%s')",
					 i,e(item,'msgid'),e(item,'msgstr'),e(item,'msgctxt'),file)),format("%s,line number:%d",item["msgstr"],i))
      end
      tm_db=tm_db+clock()
   end
   tm_db=tm_db-clock()
   res = assert(con:execute("create index 'idx' on 'tab' (id,msgid,msgstr,msgctxt,fuzzy,flag,pof)"))
   res = assert (con:commit())
   tm_db=tm_db+clock()
   print(tm_parsing,tm_db)
end

-- parse_opt(arg)
main()
